middleware/http/oauth2/oauth2_middleware.go

```diff
@@ -15,148 +15,439 @@ package oauth2
+	err := m.meta.FromMetadata(metadata, m.logger)
+	if err != nil {
+		return nil, fmt.Errorf("invalid metadata: %w", err)
+	}
+
+	// Derive the token keys
+	err = m.meta.setTokenKeys()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to derive token keys: %w", err)
 	}
 
-	forceHTTPS := utils.IsTruthy(meta.ForceHTTPS)
-	conf := &oauth2.Config{
-		ClientID:     meta.ClientID,
-		ClientSecret: meta.ClientSecret,
-		Scopes:       strings.Split(meta.Scopes, ","),
-		RedirectURL:  meta.RedirectURL,
-		Endpoint: oauth2.Endpoint{
-			AuthURL:  meta.AuthURL,
-			TokenURL: meta.TokenURL,
-		},
+	// Create the OAuth2 configuration object
+	m.meta.setOAuth2Conf()
+
+	return m.getHandler(), nil
+}
+
+// Returns the handler. This is split out to make unit testing easier.
+func (m *OAuth2Middleware) getHandler() func(next http.Handler) http.Handler {
+	// Set the callbacks depending on the mode of operation
+	switch m.meta.Mode {
+	case modeCookie:
+		m.getTokenFn = m.GetClaimsFromCookie
+		m.setTokenFn = m.cookieModeSetTokenInResponse
+		m.claimsForAuthFn = m.cookieModeClaimsForAuth
+	case modeHeader:
+		m.getTokenFn = m.headerModeGetClaimsFromHeader
+		m.setTokenFn = m.headerModeSetTokenResponse
 	}
 
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			session := sessions.Start(w, r)
+	return m.handler
+}
 
-			if session.GetString(meta.AuthHeaderName) != "" {
-				r.Header.Add(meta.AuthHeaderName, session.GetString(meta.AuthHeaderName))
-				next.ServeHTTP(w, r)
+func (m *OAuth2Middleware) handler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// If we have the "state" and "code" parameter, we need to exchange the authorization code for the access token
+		code := r.URL.Query().Get("code")
+		state := r.URL.Query().Get("state")
+		if code != "" && state != "" {
+			// Always get the claims from the cookies in this case
+			claims, err := m.GetClaimsFromCookie(r)
+			if err != nil {
+				// If the cookie is invalid, redirect to the auth endpoint again
+				// This will overwrite the old cookie
+				m.logger.Debugf("Invalid session cookie: %v", err)
+				m.redirectToAuthenticationEndpoint(w, r)
 				return
 			}
 
-			// Redirect to the auth server
-			state := r.URL.Query().Get(stateParam)
-			if state == "" {
-				id, err := uuid.NewRandom()
-				if err != nil {
-					httputils.RespondWithError(w, http.StatusInternalServerError)
-					m.logger.Errorf("Failed to generate UUID: %v", err)
-					return
-				}
-				idStr := id.String()
-
-				session.Set(savedState, idStr)
-				session.Set(redirectPath, r.URL)
-
-				url := conf.AuthCodeURL(idStr, oauth2.AccessTypeOffline)
-				httputils.RespondWithRedirect(w, http.StatusFound, url)
+			m.exchangeAccessCode(w, r, claims, code, state)
```

Comment:
Here, we rely on `code` being not blank to exchange it to get access token. Should we also check `response_type` somehow before doing so?
Ref: https://darutk.medium.com/diagrams-and-movies-of-all-the-oauth-2-0-flows-194f3c3ade85 and https://stackoverflow.com/a/74547318/18848251